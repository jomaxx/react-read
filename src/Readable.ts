const key = Symbol('read');

interface Suspendable<T> extends PromiseLike<T> {
  [key]?: () => T;
}

export class Readable<T> {
  static create<T>(object: Suspendable<T> | T) {
    return new Readable(object);
  }

  constructor(object: Suspendable<T> | T) {
    if (!isSuspendable<T>(object)) {
      this._read = () => object;
      return;
    }

    if (!object[key]) {
      const suspender = Promise.resolve(object);

      object[key] = () => {
        throw suspender;
      };

      suspender.then(
        result => {
          object[key] = () => result;
        },
        error => {
          object[key] = () => {
            throw rethrown(error);
          };
        },
      );
    }

    this._read = () => (object[key] as () => T)();
  }

  _read: () => T;

  get value() {
    return this._read();
  }

  get read() {
    return this._read;
  }
}

function isSuspendable<T>(obj: any): obj is Suspendable<T> {
  return obj && typeof obj.then === 'function';
}

function rethrown(old: Error) {
  if (!(old instanceof Error)) return old;
  const error = new Error(old.message);
  error.name = old.name;
  error.message = old.message;
  if (!error.stack || !old.stack) return old;
  const stack = error.stack.split('\n');
  stack.splice(1, 1);
  stack.push('    rethrown:');
  stack.push(...old.stack.split('\n').slice(1));
  error.stack = stack.join('\n');
  return error;
}
