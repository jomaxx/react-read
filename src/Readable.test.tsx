import React from 'react';
import ReactDOM from 'react-dom';
import { Readable } from './Readable';

beforeEach(() => {
  jest.restoreAllMocks();
});

test('returns', () => {
  return new Promise(resolve => {
    const expected = {};
    const readable = Readable.create(expected);

    function Test() {
      const { value: received } = readable;

      React.useEffect(() => {
        expect(received).toBe(expected);
        resolve();
      }, []);

      return null;
    }

    ReactDOM.render(<Test />, document.createElement('div'));
  });
});

test('resolves', () => {
  return new Promise(resolve => {
    const expected = {};
    const readable = Readable.create(Promise.resolve(expected));

    function Test() {
      const received = readable.read();

      React.useEffect(() => {
        expect(received).toBe(expected);
        resolve();
      }, []);

      return null;
    }

    ReactDOM.render(
      <React.Suspense fallback={null}>
        <Test />
      </React.Suspense>,
      document.createElement('div'),
    );
  });
});

test('rejects', () => {
  return new Promise(resolve => {
    jest.spyOn(console, 'error').mockReturnValue(undefined);

    const expected = new Error('error');
    const readable = Readable.create(Promise.reject(expected));

    class ErrorBoundary extends React.Component {
      componentDidCatch(received: Error) {
        expect(received.stack).toMatchSnapshot();
        resolve();
      }

      render() {
        return this.props.children;
      }
    }

    function Test() {
      readable.read();
      return null;
    }

    ReactDOM.render(
      <ErrorBoundary>
        <React.Suspense fallback={null}>
          <Test />
        </React.Suspense>
      </ErrorBoundary>,
      document.createElement('div'),
    );
  });
});
