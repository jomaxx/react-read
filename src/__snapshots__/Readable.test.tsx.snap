// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejects 1`] = `
"Error: error
    at Promise.object.<computed> (/Users/jomaxx/dev/jomaxx/react-read/src/Readable.ts:31:19)
    at Readable._read (/Users/jomaxx/dev/jomaxx/react-read/src/Readable.ts:37:48)
    at Test (/Users/jomaxx/dev/jomaxx/react-read/src/Readable.test.tsx:73:16)
    at renderWithHooks (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:16260:18)
    at mountIndeterminateComponent (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:18794:13)
    at beginWork$1 (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:20162:16)
    at HTMLUnknownElement.callCallback (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:336:14)
    at innerInvokeEventListeners (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:316:27)
    at invokeEventListeners (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:267:3)
    at HTMLUnknownElementImpl._dispatch (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:214:9)
    at HTMLUnknownElementImpl.dispatchEvent (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:87:17)
    at HTMLUnknownElement.dispatchEvent (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:144:23)
    at Object.invokeGuardedCallbackDev (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:385:16)
    at invokeGuardedCallback (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:440:31)
    at beginWork$$1 (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:25780:7)
    at performUnitOfWork (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:24698:12)
    at workLoopSync (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:24671:22)
    at performSyncWorkOnRoot (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:24270:11)
    at /Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:12199:24
    at unstable_runWithPriority (/Users/jomaxx/dev/jomaxx/react-read/node_modules/scheduler/cjs/scheduler.development.js:697:12)
    at runWithPriority$2 (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:12149:10)
    at flushSyncCallbackQueueImpl (/Users/jomaxx/dev/jomaxx/react-read/node_modules/react-dom/cjs/react-dom.development.js:12194:7)
    at workLoop (/Users/jomaxx/dev/jomaxx/react-read/node_modules/scheduler/cjs/scheduler.development.js:641:34)
    at flushWork (/Users/jomaxx/dev/jomaxx/react-read/node_modules/scheduler/cjs/scheduler.development.js:596:16)
    at _flushCallback (/Users/jomaxx/dev/jomaxx/react-read/node_modules/scheduler/cjs/scheduler.development.js:48:9)
    at Timeout.callback [as _onTimeout] (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jsdom/lib/jsdom/browser/Window.js:643:19)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
    rethrown:
    at /Users/jomaxx/dev/jomaxx/react-read/src/Readable.test.tsx:58:22
    at new Promise (<anonymous>)
    at Object.<anonymous> (/Users/jomaxx/dev/jomaxx/react-read/src/Readable.test.tsx:55:10)
    at Object.asyncJestTest (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:100:37)
    at /Users/jomaxx/dev/jomaxx/react-read/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (/Users/jomaxx/dev/jomaxx/react-read/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /Users/jomaxx/dev/jomaxx/react-read/node_modules/jest-jasmine2/build/queueRunner.js:75:41"
`;
